openquake/hazardlib/near_fault.py

```diff
@@ -0,0 +1,476 @@
+# The Hazard Library
+# Copyright (C) 2012-2014, GEM Foundation
+#
+# This program is free software: you can redistribute it and/or modify
+# it under the terms of the GNU Affero General Public License as
+# published by the Free Software Foundation, either version 3 of the
+# License, or (at your option) any later version.
+#
+# This program is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU Affero General Public License for more details.
+#
+# You should have received a copy of the GNU Affero General Public License
+# along with this program.  If not, see <http://www.gnu.org/licenses/>.
+"""
+Module :mod:`openquake.hazardlib.nearfault` provides methods for near fault
+PSHA calculation.
+"""
+
+import math
+import numpy as np
+from openquake.hazardlib.geo import geodetic as geod
+import scipy.spatial.distance as dst
+
+
+def get_xyz_from_ll(projected, reference):
+    """
+    This method computes the x ,y and z coordinates of a set of points
+    provided a reference point
+
+    :param projected:
+        :class:`~openquake.hazardlib.geo.point.Point` object
+        representing the coordinates of target point to be projected
+    :param reference:
+        :class:`~openquake.hazardlib.geo.point.Point` object
+        representing the coordinates of the reference point.
+
+    :returns:
+            x
+            y
+            z
+    """
+
+    azims = geod.azimuth(reference.longitude, reference.latitude,
+                         projected.longitude, projected.latitude)
+    depths = np.subtract(reference.depth, projected.depth)
+    dists = geod.geodetic_distance(reference.longitude,
+                                   reference.latitude,
+                                   projected.longitude,
+                                   projected.latitude)
+    return (dists * math.sin(math.radians(azims)),
+            dists * math.cos(math.radians(azims)),
+            depths)
+
+
+def get_plane_equation(p0, p1, p2, reference):
+    '''
+    Define plane equation by 3 given points, p0, p1, p2, which format in
+    class:`~openquake.hazardlib.geo.point.Point` object
+
+    :param p0:
+        :class:`~openquake.hazardlib.geo.point.Point` object
+        representing the location of the one vertex of the fault patch.
+    :param p1:
+        :class:`~openquake.hazardlib.geo.point.Point` object
+        representing the location of the one vertex of the fault patch.
+        Note: the order of the vertex should give
```

Comment:
I modified the description. 
The order of the three points has to follow strictly, in order to make the normal of the plane a same direction. It will be used in the average_s_rad method, and in average_s_rad, we need to define if the site is above or below the fault plane. Therefore, I need to make consistence of which normal direction is positive(need to define which direction is positive).